src/libmongoc/src/mongoc/mongoc-crypt.c

```diff
@@ -576,37 +581,158 @@ _state_need_kms (_state_machine_t *state_machine, bson_error_t *error)
 #undef BUFFER_SIZE
 }
 
+/**
+ * @brief Determine whether the given kmsProviders has an empty 'aws'
+ * subdocument
+ *
+ * @param kms_providers The user-provided kmsProviders
+ * @param error Output parameter for possible errors.
+ * @return true If 'aws' is present and an empty subdocument
+ * @return false Otherwise or on error
+ */
+static bool
+_needs_on_demand_aws_kms (bson_t const *kms_providers, bson_error_t *error)
+{
+   bson_iter_t iter;
+   if (!bson_iter_init_find (&iter, kms_providers, "aws")) {
+      // No "aws" subdocument
+      return false;
+   }
+
+   if (!BSON_ITER_HOLDS_DOCUMENT (&iter)) {
+      // "aws" is not a document? Should be validated by libmongocrypt
+      return false;
+   }
+
+   const uint8_t *dataptr;
+   uint32_t datalen;
+   bson_iter_document (&iter, &datalen, &dataptr);
+   bson_t subdoc;
+   if (!bson_init_static (&subdoc, dataptr, datalen)) {
+      // Invalid "aws" document? Should be validated by libmongocrypt
+      return false;
+   }
+
+   if (bson_empty (&subdoc)) {
+      // "aws" is present and is an empty subdocument, which means that the user
+      // requests that the AWS credentials be loaded on-demand from the
+      // environment.
+      return true;
+   } else {
+      // "aws" is present and is non-empty, which means that the user has
+      // already provided credentials for AWS.
+      return false;
+   }
+}
+
+/**
+ * @brief Attempt to load AWS credentials from the environment and insert them
+ * into the given kmsProviders bson document on the "aws" property.
+ *
+ * @param out A kmsProviders object to update
+ * @param error An error-out parameter
+ * @return true If there was no error. We MAY or MAY NOT have successfully added
+ * the "aws" property.
+ * @return false If there was an error while updating the BSON data or obtaining
+ * credentials.
+ */
+static bool
+_try_add_aws_from_env (bson_t *out, bson_error_t *error)
+{
+   // Attempt to obtain AWS credentials from the environment.
+   _mongoc_aws_credentials_t got_creds;
+   if (!_mongoc_aws_credentials_obtain (NULL, &got_creds, error)) {
+      // Failed to get credentials. This isn't necessarily an error.
+      return error->code == 0;
+   }
+
+   // We have the credentials. Give them back to libmongocrypt
+
+   bson_t aws = {0};
+   // Begin the new "aws" subdoc
+   if (!BSON_APPEND_DOCUMENT_BEGIN (out, "aws", &aws)) {
```

Comment:
Suggest converting `if (expr) { return false; }` into `BSON_ASSERT (expr);` in this function given failure is not recorded as an error.

Reply:
This always boils down to whether one would consider allocation failure to be recoverable, as that is the only condition in which these statements will fail.

Reply:
Given we do not have a `MONGOC_ERROR_OUT_OF_MEMORY` or equivalent error code to report this class of failure, I am in favor of converting these to assertions. If not converted to assertions, `error` should at least be set prior to `goto fail;` under the `// Error while trying to add AWS credentials` condition.